float32 x_value
float32 y_value
float32 z_value